### DISCUSSION TITLE
How to always have ChecksumSHA256?

### DISCUSSION THREAD

**Discussion by daddz** [minio/minio#20400] [Q&A] [answered]
Is there any way to have SHA256 checksums for all objects (in a bucket) like in S3 (https://aws.amazon.com/de/blogs/aws/new-additional-checksum-algorithms-for-amazon-s3/)?

Is there a way to add SHA256 checksums to existing objects without download/calculate/re-upload? 
I tried the "copy" approach with adding the header but that didn't work in minio: https://aws.amazon.com/de/blogs/storage/enabling-and-validating-additional-checksums-on-existing-objects-in-amazon-s3/

**Comment by harshavardhana** [comment, marked as answer]
@daddz what is the total size of your bucket? And how big is your deployment? 

Remember, doing what AWS recommends is unnecessary on MinIO. We already do bitrot verification on the persistent data every time you read the object. 

So, doing sha256 is simply a waste of time since it is prolonged and CPU-hungry unless you have crazy cores and vast CPU resources to spare.

**Reply by daddz** [reply]
It's a single instance with just 4TB of data - but that's not the issue here.

For my usecase I need the sha256 checksum and I can't guarantee it will always be provided through the various upload methods.

* `PresignedPutObject` has no option to add custom headers/control the checksum
* `PutObject` only stores the checksum (of the whole file, not checksum of checksums) if I set a custom header and disable multipart upload
* `PresignedPostPolicy` doesn't store the provided checksum through `SetChecksum`

Right now I'd need to make sure that any upload provides the checksum if possible and after every upload check if the object has a checksum and if not, download, calculate and re-upload it.

So I thought since minio does calculate/store the checksum in some cases, why not have an option to control it explicitly, e.g. per bucket.

Either way I think it would be great if the docs could somehow explain in which cases checksums are calculated and stored.

**Reply by harshavardhana** [reply]
In all cases checksum is always present on the drive at block level.

Adding sha256 on top doesn't really add any benefits.

**Reply by daddz** [reply]
> Adding sha256 on top doesn't really add any benefits.

But it is there already? Minio already calculates it in some cases. I'm merely asking if I can force it do it everytime or an option for that could be added.

**Reply by klauspost** [reply]
No, not SHA256, but internal bitrot protection. SHA256 is too slow.

**Reply by daddz** [reply]
Maybe I am completely misunderstanding what minio is doing. I tried to follow the code and figure out when it does and when it doesn't do these checksums but I don't think I have the full picture yet.

So when I do this:
```go
minioClient.PutObject(ctx, "somebucket", "someobject", someBuf, 128, minio.PutObjectOptions{
	UserMetadata:     map[string]string{"x-amz-checksum-sha256": "somebase64checksum"},
	DisableMultipart: true,
})
```

I would now assume minio is calculating the checksum, otherwise it wouldn't be able to check the file against the provided checksum.

When doing the above `PutObject` I can then later read out the SHA256 checksum via:
```go
info, _ := minioClient.StatObject(ctx, "somebucket", "someobject", minio.StatObjectOptions{Checksum: true})
fmt.Println(info.ChecksumSHA256)
```

I'm simply looking for a way to ensure that this `StatObject` call always contains that checksum because in many cases it doesn't (e.g. when uploading via `PresignedPutUrl`) and it's not really clear to me when it does and when it doesn't.

I know that SHA256 is slow but that is totally fine - this behavior shouldn't be the default but it would be great to have the option.
From the implementation it looks like all the building blocks are there already.

Maybe as a last point that could help clearing up some confusion would be adding some documentation for the `DisableContentSha256` and `AutoChecksum` fields in `PutObjectOptions`:

`DisableContentSha256` sounds like it calculates SHA256 by default and you need to disable it explicitly - I still don't know when and how this option actually affects anything

`AutoChecksum` has the following documentation:
```go
// AutoChecksum is the type of checksum that will be added if no other checksum is added,
// like MD5 or SHA256 streaming checksum, and it is feasible for the upload type.
// If none is specified CRC32C is used, since it is generally the fastest.
```
That sounds like CRC32 is always calculated but when doing `StatObject` it is always empty.
Providing any other option to this also doesn't seem to have any effect.
Also it is unclear what a "feasible upload type" is.

**Reply by klauspost** [reply]
>  and it's not really clear to me when it does and when it doesn't

If content is uploaded with a checksum, it will be verified and saved, and therefore returned on HeadObject/Stat.

DisableContentSha256 is not related to checksums. On Non-TLS connections this will sign the request content with SHA256.

> That sounds like CRC32 is always calculated

No. Read is again. "if no other checksum is added" - that could also be Content SHA, MD5..  IIRC TLS will also disable it, since TLS ensures the integrity of the transferred data.

"and it is feasible for the upload type", means the contents cannot be streaming. This depends on the input type.

`AutoChecksum` will only add checksums if it can be done without cost. In some cases it will require the client to have been created with [TrailingHeaders](https://github.com/minio/minio-go/blob/b81adb28a5f91d71877789bb19b9ffa333dfd5c9/api.go#L121).

**Reply by daddz** [reply]
> If content is uploaded with a checksum, it will be verified and saves, and therefore returned on HeadObject/Stat.

Are you open to the idea of adding a bucket option to just always calculate some/certain checksums and have it saved? 

> DisableContentSha256 is not related to checksums. On Non-TLS connections this will sign the request content with SHA256.

Ah okay that is helpful, I think adding that to the documentation would certainly be helpful for others as well.

> IIRC TLS will also disable it, since TLS ensures the integrity of the transferred data.

Is there also some special case when using localhost - e.g. minio in docker with port mapping (localhost:9000)? In that case at least it also has no checksums at all.

> AutoChecksum will only add checksums if it can be done without cost. In some cases it will require the client to have been created with [TrailingHeaders](https://github.com/minio/minio-go/blob/b81adb28a5f91d71877789bb19b9ffa333dfd5c9/api.go#L121).

Could you elaborate? 
* What cases lead to "no cost"?
* Which cases require `TrailingHeaders`?

I think adding all the information you provided would be very helpful in the code comments/documentation.

It seems like lot of the checksumming that currently exists revolves around ensuring the integrity of the transferred data whereas I'm looking into problems when working with data that is already in the system: 
* Is this object already in the bucket/minio?
* Are these 2 objects in my bucket the same (content-wise)?

For these problems I thought I could piggy-back on the work the storage system - be it S3 or minio - already has done: a checksum.

Thanks for taking the time to answer my questions!

**Reply by klauspost** [reply]
> Are you open to the idea of adding a bucket option to just always calculate some/certain checksums and have it saved?

We generally don't extend AWS S3 API, since it promotes vendor lock-in - and that is actually not a goal we have.
If there is a commercial request for it we might prioritize it.

> Is there also some special case when using localhost

No. You can follow the code paths from here: https://github.com/minio/minio-go/blob/48517d86c24736ce70e167d6e038e02486ad1872/api-put-object.go#L302

For single part PutObject calls this is the decision code: https://github.com/minio/minio-go/blob/21381fc6dd612fd7f1f03b1e5d5fddef4b6eae52/api-put-object-streaming.go#L750-L759

> Could you elaborate?

Follow the code. Listing out every condition is way beyond what I can commit to here. There are checksums added when it makes sense. There is currently no mechanism for adding them fully automatic. Thus it is not documented.
